api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+	for _, txn := range txns {
+		s.wait(txn)
+	}
+	return nil
+}
+
+func (s *Service) bindTicker(ctx context.Context) {
+	go func() {
+		for {
+			<-s.ticker.C
+			if err := s.waitAllQualifying(ctx, false); err != nil {
+				log.Errorf("waitAllQualifying from ticker: %v", err)
+			}
+		}
+	}()
+}
+
+func (s *Service) SendFil(ctx context.Context, req *pb.SendFilRequest) (*pb.SendFilResponse, error) {
+	f, err := address.NewFromString(req.From)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing from address: %v", err)
+	}
+	t, err := address.NewFromString(req.To)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing to address: %v", err)
+	}
+	nanoAmount := (&big.Int{}).SetInt64(req.AmountNanoFil)
+	factor := (&big.Int{}).SetInt64(int64(math.Pow10(9)))
+	amount := (&big.Int{}).Mul(nanoAmount, factor)
+	msg := &types.Message{
+		From:  f,
+		To:    t,
+		Value: types.BigInt{Int: amount},
+	}
+	client, cls, err := s.clientBuilder(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "creating filecoin client: %v", err)
+	}
+	defer cls()
+
+	sm, err := client.MpoolPushMessage(ctx, msg, nil)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "pushing message: %v", err)
+	}
+
+	now := time.Now()
+
+	tx := txn{
+		ID:            primitive.NewObjectID(),
+		From:          req.From,
+		To:            req.To,
+		AmountNanoFil: req.AmountNanoFil,
+		MessageCids:   []msgCid{{Cid: sm.Message.Cid().String(), CreatedAt: now}},
+		MessageState:  pb.MessageState_MESSAGE_STATE_PENDING,
+		CreatedAt:     now,
+		UpdatedAt:     now,
+	}
+
+	if _, err = s.col.InsertOne(ctx, &tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "inserting txn into collection: %v", err)
+	}
+
+	wait := s.wait(tx)
+
+	if req.Wait {
+		res := <-wait
+		if res.err != nil {
+			return nil, status.Errorf(codes.Internal, "waiting for result: %v", res.err)
+		}
+		tx = res.txn
+	}
+
+	pbTx, err := toPbTxn(tx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "converting to pb txn: %v", err)
+	}
+
+	return &pb.SendFilResponse{Txn: pbTx}, nil
+}
+
+func (s *Service) Txn(ctx context.Context, req *pb.TxnRequest) (*pb.TxnResponse, error) {
+	res := s.col.FindOne(ctx, bson.M{"message_cids.cid": req.MessageCid})
+	if res.Err() == mongo.ErrNoDocuments {
+		return nil, status.Error(codes.NotFound, "no txn found for cid")
+	}
+	if res.Err() != nil {
+		return nil, status.Errorf(codes.Internal, "querying for cid txn: %v", res.Err())
+	}
+
+	var tx txn
+	if err := res.Decode(&tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding cid txn result: %v", res.Err())
+	}
+
+	if tx.MessageState != pb.MessageState_MESSAGE_STATE_ACTIVE && tx.MessageState != pb.MessageState_MESSAGE_STATE_FAILED && req.Wait {
+		res := <-s.wait(tx)
+		if res.err != nil {
+			return nil, status.Errorf(codes.Internal, "waiting for result: %v", res.err)
+		}
+		tx = res.txn
```

Comment:
If requested, and the txn isn't already in a final state, we wait for the message to be active on chain (or time out or fail). Similar to `SendFil` in this way.